contents/blog/null-handling-hogql.md

```diff
@@ -0,0 +1,82 @@
+
+We recently had challenges with null during the [release of HogQL](/blog/introducing-hogql). Our engineers Marius and Michael realized null values for properties were leading to inaccurate results and customer confusion. This post covers their efforts to figure out and implement a solution for this.
+
+## The problem with null in HogQL
+
+The issue with null in HogQL was that SQL is strict about nulls. If any value in an expression is null, then the entire expression is null. Here are two HogQL expressions showcasing why this is a problem:
+
+1. `sumIf(properties.money, event = 'no problems') + sumIf(properties.money - 10, event = 'some problems')` returns null if either of the sides of the equation are null. This  happens if there are no matching events (named `no problems` or `some problems`) or the property (`money`) is null.
+2. `concat('money: ', properties.money)` returns null if there's no `money` property, though users might expect either `'money: '`  or `'money: null'`
+
+PostHog’s existing solutions for handling null did not work with the full access HogQL provides. Specifically:
+
+- ClickHouse skips null during aggregation, but users could want to concat or check null values with HogQL.
+- PostHog previously treated nulls as strings, but this could include null values in counts when they shouldn’t be.
+
+At this stage in HogQL’s development, if a value could be null, users must handle it themselves. For example, users must wrap each `sumIf` function in the first example above in a `nullIf(x, 0)` like this:
+
+```
+nullIf(
+	sumIf(properties.money, event = 'no problems'), 
+	0
+) + nullIf(
+	sumIf(properties.money - 10, event = 'some problems'),
+	0
+)
+```
+
+This is not a good user experience as it is non-obvious they must do this and leads to a lot of extra code and checks.
+
+## Brainstorming potential solutions
+
+To solve the issue with nulls, Marius and Michael came up with a bunch of potential solutions, (none of which were inevitably chosen):
+
+1. **Returning a default value depending on type.** For example, returning `""` for strings, `0` for numbers, or `false` for booleans. This didn’t handle the expression `properties.bla is null` and required rewriting `is set` and `is not set` logic.
+
+2. **Including a nullable checkbox in data management.** Similar to the previous option, but required users to input if they expect a property to be null. This required a lot of user work and remains confusing if they didn’t understand how to handle null.
+
+3. **Create JS-like `'bla' (NOT) IN properties` expression logic.** We felt this was too complex for users and makes the question of `is null` for an unset property unclear. 
+
+4. **Use the ClickHouse function `sumOrDefaultIf`.** Faces the same problem of being null if any values are null.
```

Comment:
This is not exactly true. We will likely expose all the aggregations variations like `sum` (exposed), `sumIf` (exposed), `sumOrDefault`, `sumOrDefaultIf`, and perhaps some more. TODO

Also, it'll be null if _all_ values are nulls 🤷 

```sql
select sumOrDefault(a) from (select arrayJoin([1,2,3, null]) as a) -- returns 6
select sumOrDefault(a) from (select arrayJoin([null, null, null]) as a) -- returns null, NOT 1
```

Also,

```sql
select avg(a) from (select arrayJoin([1,2,3, null]) as a) -- returns 2 as nulls are just ignored
```